initrd/etc/ash_functions

```diff
@@ -53,6 +53,154 @@ preserve_rom() {
+		message="Please confirm that your GPG card is inserted(Y/n) or your GPG key material (b)backup thumbdrive is inserted [Y/n/b]: "
+	else
+		# Generic message if no known key material backup
+		message="Please confirm that your GPG card is inserted [Y/n]: "
+	fi
+
+	read \
+		-n 1 \
+		-p "$message" \
+		card_confirm
+	echo
+
+	if [ "$card_confirm" != "y" \
+		-a "$card_confirm" != "Y" \
+		-a "$card_confirm" != "b" \
+		-a -n "$card_confirm" ] \
+		; then
+		die "gpg card not confirmed"
+	fi
+
+	# If user has known GPG key material Thumb drive backup and asked to use it
+	if [[ "$CONFIG_HAVE_GPG_KEY_BACKUP" == "y" && "$card_confirm" == "b" ]]; then
+		#Only mount and import GPG key material thumb drive backup once
+		if [ ! "$CONFIG_GPG_KEY_BACKUP_IN_USE" == "y" ]; then
+			CR_NONCE="/tmp/secret/cr_nonce"
+			CR_SIG="$CR_NONCE.sig"
+
+			#Wipe any previous CR_NONCE and CR_SIG
+			shred -n 10 -z -u "$CR_NONCE" "$CR_SIG" >/dev/null 2>&1 || true
+
+			#Prompt user for provisioned GPG Admin PIN that will be passed along to mount-usb and to import gpg subkeys
+			echo
+			#TODO: change all passphrase prompts in codebase to include -r to prevent backslash escapes
+			read -r -s -p "Please enter GPG Admin PIN needed to use the GPG backup thumb drive: " gpg_admin_pin
+			echo
+			#prompt user to select the proper encrypted partition, which should the first one on next prompt
+			warn "Please select encrypted LUKS on GPG key material backup thumb drive (not public labeled one)"
+			mount-usb --pass "$gpg_admin_pin" || die "Unable to mount USB with GPG Admin PIN"
+			echo "++++ Testing detach-sign operation and verifiying against fused public key in ROM"
+			gpg --pinentry-mode=loopback --passphrase-file <(echo -n "${gpg_admin_pin}") --import /media/subkeys.sec >/dev/null 2>&1 ||
+				die "Unable to import GPG private subkeys"
+			#Do a detach signature to ensure gpg material is usable and cache passphrase to sign /boot from caller functions
+			dd if=/dev/urandom of="$CR_NONCE" bs=20 count=1 >/dev/null 2>&1 ||
+				die "Unable to create $CR_NONCE to be detach-signed with GPG private signing subkey"
+			gpg --pinentry-mode=loopback --passphrase-file <(echo -n "${gpg_admin_pin}") --detach-sign "$CR_NONCE" >/dev/null 2>&1 ||
+				die "Unable to detach-sign $CR_NONCE with GPG private signing subkey using GPG Admin PIN"
+			#verify detached signature against public key in rom
+			gpg --verify "$CR_SIG" "$CR_NONCE" > /dev/null 2>&1 && \
+                                echo "++++ Imported private subkeys match public key fused in rom and can be used under Heads"  || \
+			 	die "Unable to verify $CR_SIG detached signature against public key in ROM"
+			#Wipe any previous CR_NONCE and CR_SIG
+			shred -n 10 -z -u "$CR_NONCE" "$CR_SIG" >/dev/null 2>&1 || true
+			#TODO: maybe just an export instead of setting /etc/user.config otherwise could be flashed in weird corner case situation
+			set_user_config "CONFIG_GPG_KEY_BACKUP_IN_USE" "y"
```

Comment:
@JonathonHall-Purism this is another place I would love to have insights from you. By toggling this, I would have had expected recovery shell, called after oem-factory-reset complettion (now skipping flash on qemu board) to prompt the user through gpg_auth. But it doesn't. Also, env doesn't show CONFIG_GPG_KEY_BACKUP_IN_USE so I'm clearly something

Reply:
How are you getting from oem-factory-reset to the recovery shell?  oem-factory-reset reboots, so if you reboot the config is lost in qemu since we can't flash it.

There might be some ways to stitch those steps together in qemu but I'm not sure it's worth it, sound pretty complex for a test platform to work around the inability to flash anyway :thinking:

Reply:
@JonathonHall-Purism in current state of PR, oem-factory-reset skips flashing and proposes to go to recovery shell when in debug mode, recovery sources . /tmp/config and calls gpg_auth. Considering that codepath, I would have expected gpg_auth to be called, but it seems that at that point, . /tmp/config being sourced doesn't contain CONFIG_GPG_KEY_BACKUP_IN_USE=y and i'm not sure why, since set_user_config "CONFIG_GPG_KEY_BACKUP_IN_USE" "y" sould be set correctly, combined and sourced from . /tmp/config as well?

Reply:
Commented elsewhere, I noticed that CONFIG_GPG_KEY_BACKUP_IN_USE is only set in the RSA key logic, not the p256 key logic.  Most of that logic doesn't depend on GPG_ALGO so it seems like it should be moved out of the if-else.  Were you by chance generating a p256 key?